sandbox/mooncake.py

```diff
@@ -0,0 +1,252 @@
+            num_channels, num_channels, kernel_size=1, dilation=1)
+
+    def forward(self, x):
+        a = self.conv1(x)
+        b = torch.tanh(a[:, :self.num_channels, :])
+        c = torch.sigmoid(a[:, self.num_channels:, :])
+
+        return self.conv2(b * c) + x
+
+
+class Mooncake(nn.Module):
+    # TODO tweak parameters
+    def __init__(
+            self,
+            in_channels=1,
+            num_channels=4,
+            num_blocks=8,
+            kernel_size=2,
+            dilations=[1, 2, 4, 8, 16, 32, 64, 128],
+            num_components=5,
+    ):
+        super(Mooncake, self).__init__()
+        self.num_channels = num_channels
+        self.num_blocks = num_blocks
+
+        self.conv1 = CausalConv1d(
+            in_channels, num_channels, kernel_size=kernel_size, dilation=1)
+        self.conv2 = nn.Conv1d(num_channels, 2 * num_channels, kernel_size=1)
+
+        self.blocks = nn.ModuleList([
+            ResidualBlock(
+                num_channels, kernel_size=kernel_size,
+                dilation=dilations[i]).to(DEVICE)
+            for i in range(self.num_blocks)
+        ])
+
+        self.conv_pi = nn.Conv1d(
+            2 * num_channels, num_components + 1, kernel_size=1, bias=False)
+        self.conv_mu = nn.Conv1d(
+            2 * num_channels, num_components, kernel_size=1, bias=False)
+        self.conv_sigma = nn.Conv1d(
+            2 * num_channels, num_components, kernel_size=1, bias=False)
+
+    def forward(self, x):
+        """
+        Parameters
+        ----------
+        x : [batch_size, num_samples]
+            Input.
+
+        Returns
+        -------
+        pi, mu, sigma : [batch_size, num_components (+1 for pi), num_samples]
+            Mixture weights, means and standard deviations of each Gaussian
+            component, respectively.
+        """
+        x = x.unsqueeze(1)  # Sinewave only has 1 channel
+        x = self.conv1(x)
+
+        residual = x
+        for i in range(self.num_blocks):
+            residual = self.blocks[i](residual)
+            x += residual
+
+        z = F.relu(x)
+        z = self.conv2(z)
+
+        pi = F.softmax(self.conv_pi(z), dim=1)
+        mu = torch.tanh(self.conv_mu(z))
+        sigma = F.softplus(self.conv_sigma(z))
+
+        return pi, mu, sigma
+
+
+if __name__ == "__main__":
+    torch.set_default_tensor_type(torch.FloatTensor)
+    np.random.seed(1618)
+    torch.manual_seed(1618)
+
+    batch_size = 4096
+    max_num_samples = 200
+
+    mooncake = Mooncake().to(DEVICE)
+    optimizer = optim.Adam(mooncake.parameters(), lr=0.001)
+
+    # Train
+    mooncake.train()
+
+    for i in range(NUM_TRAIN_ITERATIONS):
+        # Generate one sinewave and make a batch out of it.
+        sinewaves = generate_sinewaves(1, max_num_samples=max_num_samples)
+        pi, mu, sigma = mooncake(sinewaves)
+
+        # Update
+        optimizer.zero_grad()
+        nlogp = negative_log_prob(sinewaves, pi, mu, sigma).sum(dim=1).mean()
```

Comment:
just change .sum(dim=1).mean to .mean. it plays better with adam even if it's not technically "correct"

Reply:
@ccurro can you explain why it plays better with adam?

Reply:
Actually adam is scale invariant. So disregard the later part of my comment.

It'll be nice to report the mean here since we'll be reporting average likelihood score per sample regardless of how many samples an example takes up. We don't want examples with different lengths to have different likelihood scores just b/c they're different lengths - that makes it harder to compare for us.